### ISSUE TITLE
Gradients through nbest lattice?

### ISSUE THREAD

**Issue by un-baised** [k2-fsa/k2#1061]
Hi,

I'm trying to implement MWER loss in k2. Is there an easy way to get gradients through nbest decoding? Or have you implemented something similar in the past?

**Comment by danpovey** [comment]
The gradients should propagate from the original lattice to the sampled lattice you get from nbest, so you could compute an expected loss over a weighted sum of the nbest paths (weighted in a way involving the scores) and gradients would be propagated correctly.

**Comment by un-baised** [comment]
Thanks for your note, now got gradients back to the initial lattice scores. Now looking at getting gradients through k2.RnntDecodingStreams .advance back to the joiner output log probabilities but not having much luck yet. Do you have any recommendations for setting this up? I can post a code snippet if that would help.

**Comment by francisr** [comment]
I'd also like to know if it's possible to propagate lattice gradients to the frame log probs through the decoder.

**Comment by danpovey** [comment]
I missed this conversation.  @pkufool does the RnntDecodingStreams object support gradients?

**Comment by francisr** [comment]
I had a quick look at the cuda code. I'm not very familiar with it, but from what I understand it's not supported.  
Would it be complicated to implement?

**Comment by csukuangfj** [comment]
I think it will return the arc indexes , so it should be possible to support gradient back propagation.

**Comment by csukuangfj** [comment]
> I think it will return the arc indexes , so it should be possible to support gradient back propagation.

https://github.com/k2-fsa/k2/blob/1b151e5280046757a118fd1e9a5ed9dde8d18345/k2/python/k2/rnnt_decode.py#L182-L184

https://github.com/k2-fsa/k2/blob/1b151e5280046757a118fd1e9a5ed9dde8d18345/k2/csrc/rnnt_decode.h#L191-L199

You can see that it returns the arc_map.

**Comment by danpovey** [comment]
Do one of you guys have time to implement this, or shoud I try to figure out how?

**Comment by danpovey** [comment]
Oh.  I thought perhaps it would be outputting the scores somehow, e.g. into some kind of lattice-like object.  Yes, unless it is for the scores it doesn't make sense.

**Comment by francisr** [comment]
We do get the lattice out from the decoder, whose scores will come from the log probs we feed in `.advance()`.

**Comment by francisr** [comment]
The k2 code is quite cryptic to me right now (especially the CUDA stuff), so it would take me a bit too long.

**Comment by csukuangfj** [comment]
You can implement it purely in Python and it has nothing to do with CUDA.

https://github.com/k2-fsa/k2/blob/1b151e5280046757a118fd1e9a5ed9dde8d18345/k2/python/k2/rnnt_decode.py#L33
It takes an `Fsa` as input and it also produces an `Fsa` as output:
https://github.com/k2-fsa/k2/blob/1b151e5280046757a118fd1e9a5ed9dde8d18345/k2/python/k2/rnnt_decode.py#L153

We also have the arc map specifying the relationship between the input `Fsa` and the output `Fsa`:
https://github.com/k2-fsa/k2/blob/1b151e5280046757a118fd1e9a5ed9dde8d18345/k2/python/k2/rnnt_decode.py#L182

Basically, you can follow `k2.arc_sort()` or `k2_shortest_path()` to implement the backward pass.

**Comment by francisr** [comment]
But this gives the link between the decode graph scores and the lattice scores. I need to link the neural network logprobs from each frame to the lattice scores.

**Comment by csukuangfj** [comment]
> But this gives the link between the decode graph scores and the lattice scores. I need to link the neural network logprobs from each frame to the lattice scores.

Ah, I see.  If you need the gradient for nbest list, that would be much simpler and can be done in python.
It is more difficult for the whole lattice.

---

For the nbest list case, let us consider one path.  The length of the path equals to the number of frames in `log_probs`.
The i-th arc corresponds to the i-th frame and the label `k` on the i-th arc corresponds to `log_probs[i][k]` . You can
use this relationship to backprogate gradient from the nbest graph to `log_probs`.

**Comment by francisr** [comment]
Not quite as easy, as during decode we have multiple log_probs at each time step, one for each context.  
We can compute which context is used based on the input labels of a batch, and then match it to the context that corresponds to a log prob, and I've implemented something like that, but I didn't find an easy way to make this a batched operation, so it's very inefficient and way too slow for training.

**Comment by danpovey** [comment]
It should be easy as long as the output scores from each decoding step are
in a torch tensor, to support grad.
Will increase memory use though, if we require grad.

**Comment by danpovey** [comment]
... easiest way would be to let the decoding step output  ints that reflect
the source of the scores in the nnet output (e.g. flattened, so just 1 int
per score), and recreate the scores using a torch indexing operation, like
we do for fsa operation scores IIRC.

**Comment by francisr** [comment]
Is it something you'd have space to do in the coming weeks?

**Comment by danpovey** [comment]
Yes I'm sure one of us will have time.
